### ISSUE TITLE
Cannot deduce type of function that should have type `int option option -> int`

### ISSUE THREAD

**Issue by julianhyde** [hydromatic/morel#126]
Cannot deduce type of a function that should have type `int option option -> int`:
```
fun f (SOME (SOME i)) = i + 1
  | f (SOME NONE) = 0
  | f NONE = ~1
```
throws
```
java.lang.RuntimeException: Cannot deduce type: conflict: option(tuple(int, T9)) vs tuple(int, T10)
	at net.hydromatic.morel.compile.TypeResolver.deduceType_(TypeResolver.java:144)
	at net.hydromatic.morel.compile.TypeResolver.deduceType(TypeResolver.java:103)
	at net.hydromatic.morel.compile.Compiles.validateExpression(Compiles.java:57)
```

**Comment by julianhyde** [comment]
Similar:
```
fun f ((SOME i) :: NIL) = i + 1
  | f NIL = 0
```
and an equivalent function, written a different way:
```
fun f [(SOME i)] = i
  | f [] = 0
```
both give `Cannot deduce type: conflict: tuple(int, T7) vs int` but should give
```
val f = fn : int option list -> int
```

**Comment by julianhyde** [comment]
Something similarly fishy going on here:
```
    final String ml11 = ""
        + "fun f ((SOME i), true) = i\n"
        + "  | f ((SOME i), false) = ~i";
    ml(ml11)
        .assertMatchCoverage(NON_EXHAUSTIVE);
```
throws
```
java.lang.ClassCastException: class net.hydromatic.morel.type.TupleType cannot be cast to class net.hydromatic.morel.type.PrimitiveType
	at net.hydromatic.morel.eval.Codes.lambda$static$0(Codes.java:265)
	at net.hydromatic.morel.compile.Inliner.visit(Inliner.java:91)
```

**Comment by julianhyde** [comment]
Update. Currently (I suspect due to the fix for #203) things have improved a little. Fragment 1 still fails, but with a different error ("no match redundant"); 2 and 3 now succeed.
```
- fun f (SOME (SOME i)) = i + 1
=   | f (SOME NONE) = 0
=   | f NONE = ~1 ;
match redundant

- fun f ((SOME i) :: NIL) = i + 1
=   | f NIL = 0;
val f = fn : int option list -> int
- f [SOME 10];
val it = 11 : int
- f [];
val it = 0 : int

- fun f [(SOME i)] = i
=   | f [] = 0;
val f = fn : int option list -> int
```

Fragment 4 seems to be mainly about the `~` operator (overloaded between `int` and `real`). It succeeds when we convert `~i` to `i + 1` (so that it is clear that we want `int`):
```
- fun f ((SOME i), true) = i
=   | f ((SOME i), false) = ~i;
java.lang.ClassCastException: class net.hydromatic.morel.type.TypeVar cannot be cast to class net.hydromatic.morel.type.PrimitiveType

- fun f ((SOME i), true) = i
=   | f ((SOME i), false) = i + 1;
val f = fn : int option * bool -> int
- f (SOME 3, true);
val it = 3 : int
- f (SOME 3, false);
val it = 4 : int
```

**Comment by julianhyde** [comment]
Fixed (presumably by #203); logged #205 for the remaining problem in fragment 1.
